### PR TITLE
Adding Multi-GPU Data Parallel Example

### PR DESCRIPTION
**Pull request by bschifferer**
It requires following PR:https://github.com/NVIDIA-Merlin/models/pull/825

### CODE REVIEW

File: examples/usecases/multi-gpu-data-parallel-training.ipynb [hunk 1]
```diff
@@ -0,0 +1,459 @@
+{
```
Comment (rnyak) [inline examples/usecases/multi-gpu-data-parallel-training.ipynb:1]: <p>typo in it's --&gt; "its batch" is the correct one.</p><p> </p><p>"All model parameters fits on a single GPU. do you mean " in this example we assume that model parameters fits on a single GPU." ?  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/855/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (rnyak) [inline examples/usecases/multi-gpu-data-parallel-training.ipynb:1]: <p>may  be rephrase: <code> import required libraries and set some hyperparameters.</code> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/855/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (rnyak) [inline examples/usecases/multi-gpu-data-parallel-training.ipynb:1]: <p>may be rephrase:  <code>We will use the same dataset and preprocessing steps with NVTabular</code> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/855/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (rnyak) [inline examples/usecases/multi-gpu-data-parallel-training.ipynb:1]: <p>this sentence is not clear "<strong>The worker with more batches waits for the gradients from the other worker, but it finished the training run.".</strong></p><p> </p><p>which worker finished the training run? the one with more batches or less batches?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/855/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (rnyak) [inline examples/usecases/multi-gpu-data-parallel-training.ipynb:1]: <p>may be rephrase as "We can take a look what required code changes are automatically applied by Merlin Models:"</p><p> </p><p>who is doing  scaling the learning rate by number of workers? where and how it is happening? if this is happening under the hood can you please refer to the code base? </p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/855/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (rnyak) [inline examples/usecases/multi-gpu-data-parallel-training.ipynb:1]: <p>how do we know we speed up the training? can you do a comparison?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/855/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (rnyak) [inline examples/usecases/multi-gpu-data-parallel-training.ipynb:1]: <p>rephrase as " if you want to learn about  how to scale NVTabular feature engineering workflows to multi-GPU , check out our <a href="https://github.com/NVIDIA-Merlin/NVTabular/blob/main/examples/03-Running-on-multiple-GPUs-or-on-CPU.ipynb" target="_blank">example</a>."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/855/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (bschifferer) [reply examples/usecases/multi-gpu-data-parallel-training.ipynb:1]: <p>Correct spelling</p><p>and updating describption</p>

Comment (bschifferer) [reply examples/usecases/multi-gpu-data-parallel-training.ipynb:1]: <p>You can compare the training time.</p><p> </p><p>A benchmark of multi-gpu training is outside the scope for an example. It should be done as part of the implementation to test the efficiency of implementation. I added it to the RMP ticket a few weeks ago.</p>

Comment (gabrielspmoreira) [reply examples/usecases/multi-gpu-data-parallel-training.ipynb:1]: <p>It is mentioned that <code>hvd.init()</code> is necessary, but could not find it in the script below.</p>

File: examples/usecases/multi-gpu-data-parallel-training.ipynb [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
```
Comment (rnyak) [inline examples/usecases/multi-gpu-data-parallel-training.ipynb:1]: <p> take a look on --&gt; take a look at</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/855/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (rnyak) [inline examples/usecases/multi-gpu-data-parallel-training.ipynb:1]: <p>take a look on --&gt; take a look at.</p><p> </p><p>what'd happen if we have a super large dataset and only 2 gpus? how we would set <code>out_files_per_proc</code>  arg then?  can it be multiply of the num_gpus like 4, 6 etc?  the reason I ask this, if we have a large set, we might want to output several parquet files.. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/855/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (rnyak) [inline examples/usecases/multi-gpu-data-parallel-training.ipynb:1]: <p>where is <code>hvd.init()</code> command? is it happening under the hood? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/855/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (rnyak) [inline examples/usecases/multi-gpu-data-parallel-training.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #5.    MPI_SIZE = int(os.getenv("OMPI_COMM_WORLD_SIZE"))
</pre><p> I get error when I type this in a notebook.. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/855/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (bschifferer) [reply examples/usecases/multi-gpu-data-parallel-training.ipynb:1]: <p>what'd happen if we have a super large dataset and only 2 gpus? how we would set&nbsp;<code>out_files_per_proc</code>&nbsp;arg then? can it be multiply of the num_gpus like 4, 6 etc? the reason I ask this, if we have a large set, we might want to output several parquet files..</p><p> </p><p>That is correct</p>

Comment (bschifferer) [reply examples/usecases/multi-gpu-data-parallel-training.ipynb:1]: <p>Yes it happens under the hood</p>

**Review by rnyak — APPROVED** [review]

**Comment by bschifferer** [comment]
rerun test

**Comment by bschifferer** [comment]
rerun tests

**Comment by bschifferer** [comment]
rerun tests
